sky/engine/bindings/jni/dart_jni.cc

```diff
@@ -112,40 +188,133 @@ JniClass::JniClass(JNIEnv* env, jclass clazz)
 JniClass::~JniClass() {
 }
 
-PassRefPtr<JniClass> JniClass::fromName(const char* name) {
-  JNIEnv* env = base::android::AttachCurrentThread();
+PassRefPtr<JniClass> JniClass::FromName(const char* name) {
+  Dart_Handle exception = NULL;
+  {
+    ENTER_JNI();
 
-  ScopedJavaLocalRef<jclass> clazz = DartJni::GetClass(env, name);
-  CheckJniException(env);
+    ScopedJavaLocalRef<jclass> clazz = DartJni::GetClass(env, name);
+    if (CheckJniException(env, &exception)) goto fail;
 
-  return adoptRef(new JniClass(env, clazz.obj()));
+    return adoptRef(new JniClass(env, clazz.obj()));
+  }
+fail:
+  Dart_ThrowException(exception);
+  ASSERT_NOT_REACHED();
 }
 
-intptr_t JniClass::getFieldId(const char* name, const char* sig) {
-  JNIEnv* env = base::android::AttachCurrentThread();
-  jfieldID id = env->GetFieldID(clazz_.obj(), name, sig);
-  CheckJniException(env);
-  return reinterpret_cast<intptr_t>(id);
+intptr_t JniClass::GetFieldId(const char* name, const char* sig) {
+  Dart_Handle exception = NULL;
+  {
+    ENTER_JNI();
+
+    jfieldID id = env->GetFieldID(clazz_.obj(), name, sig);
+    if (CheckJniException(env, &exception)) goto fail;
+
+    return reinterpret_cast<intptr_t>(id);
+  }
+fail:
+  Dart_ThrowException(exception);
+  ASSERT_NOT_REACHED();
```

Comment:
> Dart_ThrowException will transfer control to a Dart exception handler. Any code after Dart_ThrowException will not be executed.

Did you test that out?  That doesn't match my understanding of Dart_ThrowException, but I could be wrong.  My understanding was that it sets some state inside the VM and then returns control as normal.  C++ continue to execute as normal, up to the point where we return to the VM.  At that point, the VM checks whether an exception has been set by Dart_ThrowException and, if so, transfers control to the dart exception handler at that point.